d2/src/main/java/com/linkedin/d2/balancer/strategies/ClientSelector.java

```diff
@@ -0,0 +1,51 @@
+/*
+   Copyright (c) 2020 LinkedIn Corp.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+package com.linkedin.d2.balancer.strategies;
+
+import com.linkedin.d2.balancer.clients.TrackerClient;
+import com.linkedin.d2.balancer.util.hashing.HashFunction;
+import com.linkedin.d2.balancer.util.hashing.Ring;
+import com.linkedin.r2.message.Request;
+import com.linkedin.r2.message.RequestContext;
+import java.net.URI;
+import java.util.Map;
+import javax.annotation.Nullable;
+
+
+/**
+ * The selector that selects a {@link TrackerClient} from the ring to route the request to
+ */
+public interface ClientSelector
```

Comment:
This interface really is LoadBalancerStrategy was targeted to do. Maybe we can combine them?

Reply:
We have StateUpdater and ClientSelector interfaces to indicate there are 2 different steps in the load balancer strategy, so if there is one more version of strategy, we expect them to follow the same pattern to do host selection.
@zhang-chris what do you think?

Reply:
StateUpdater is really implementation details. From user interface perspective only clientSelector matters. If you want to keep the simplicity of clientSelector, how about let LoadbalanceStrategy extends clientSelector?

Reply:
In that sense, I'd rather remove ClientSelector and StateUpdater as interface. I don't think strategy should extend clientSelector. If one strategy extends client selector, we should let all strategies extend client selector.
If we don't think StateUpdater will ever expose public methods to other mps, I'm fine to remove it. @zhang-chris what do you think.

Reply:
Removed the ClientSelector and StateUpdater interface.

Reply:
Sorry, really did not mean to get rid of it.

Why let loadBalanceStrategy extend ClientSelector a bad idea? All existing strategies should still just need to implement LoadBalancerStrategy...

Reply:
Because the ClientSelector is invoked by RelativeLoadBalancerStrategy.class, client selector is used in relativeLoadBalancer to find a host from the ring after the ring is updated. StateUpdater and ClientSelector are 2 steps in the new strategy.
Because I see the interface of client selector brings the confusion, and it's more internal to the strategy implementation, so I removed the interface.